File: app_mobile/kmp/androidApp/src/main/java/com/ansmi/gestsquadre/kmp/data/SessionStorage.kt
package com.ansmi.gestsquadre.kmp.data

import android.content.Context
import com.ansmi.gestsquadre.shared.model.SquadSession
import kotlinx.datetime.Instant

/**
 * Stesso formato di [gest_squadre_session_json] nell'app Flutter 1.0.12.
 */
class SessionStorage(
    context: Context,
) {
    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun save(session: SquadSession?) {
        if (session == null) {
            prefs.edit().remove(KEY).apply()
            return
        }
        prefs.edit()
            .putString(
                KEY,
                listOf(
                    session.sessionId,
                    session.eventId,
                    session.squadId,
                    session.squadCode,
                    session.squadName,
                    session.loginAt.toString(),
                ).joinToString("|"),
            ).apply()
    }

    fun loadSessionId(): String? {
        val raw = prefs.getString(KEY, null) ?: return null
        val sessionId = raw.split("|").firstOrNull()?.trim().orEmpty()
        return sessionId.ifEmpty { null }
    }

    fun loadCached(): SquadSession? {
        val raw = prefs.getString(KEY, null) ?: return null
        val parts = raw.split("|")
        if (parts.size < 6) {
            return null
        }
        return try {
            SquadSession(
                sessionId = parts[0],
                eventId = parts[1],
                squadId = parts[2],
                squadCode = parts[3],
                squadName = parts[4],
                loginAt = Instant.parse(parts[5]),
            )
        } catch (_: Exception) {
            null
        }
    }

    fun clear() {
        prefs.edit().remove(KEY).apply()
    }

    /** Una tantum: dopo aggiornamento con log auth, forza nuovo login. */
    fun isAuthLogMigrationDone(): Boolean =
        prefs.getBoolean(KEY_AUTH_LOG_MIGRATION, false)

    fun setAuthLogMigrationDone() {
        prefs.edit().putBoolean(KEY_AUTH_LOG_MIGRATION, true).apply()
    }

    companion object {
        const val PREFS_NAME = "gest_squadre_prefs"
        const val KEY = "gest_squadre_session_json"
        private const val KEY_AUTH_LOG_MIGRATION = "kmp_auth_log_migration_v1"
    }
}

File: app_mobile/kmp/iosApp/iosApp/SessionStorage.swift
import Foundation
import shared

/// Stesso formato pipe-separated di Android `SessionStorage`.
final class SessionStorage {
    private let defaults = UserDefaults.standard
    private let key = "gest_squadre_session_json"
    private let authMigrationKey = "kmp_auth_log_migration_v1"

    func save(_ session: SquadSession?) {
        guard let session else {
            defaults.removeObject(forKey: key)
            return
        }
        let raw = [
            session.sessionId,
            session.eventId,
            session.squadId,
            session.squadCode,
            session.squadName,
            String(describing: session.loginAt),
        ].joined(separator: "|")
        defaults.set(raw, forKey: key)
    }

    func loadSessionId() -> String? {
        guard let raw = defaults.string(forKey: key) else { return nil }
        let sessionId = raw.split(separator: "|", maxSplits: 1).first.map(String.init)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard let sessionId, !sessionId.isEmpty else { return nil }
        return sessionId
    }

    func clear() {
        defaults.removeObject(forKey: key)
    }

    func isAuthLogMigrationDone() -> Bool {
        defaults.bool(forKey: authMigrationKey)
    }

    func setAuthLogMigrationDone() {
        defaults.set(true, forKey: authMigrationKey)
    }
}
